File: programmers/lv3/베스트앨범.py
"""
    lv.3 베스트앨범
    
    1.  조건 1과 2에서 제시하는 장르 별 총 재생 수 / 장르 내 각각의 재생 수 정보를 가지는 딕셔너리를 분리해 저장.
    2.  해당 딕셔너리들을 내림차순 정렬해 값을 가져와 반환.
"""


def solution(genres, plays):
    answer = []

    totalPlays = {}
    genreNumber = {}

    for i, (genre, play) in enumerate(zip(genres, plays)):
        totalPlays[genre] = totalPlays.get(genre, 0) + play

        if genre not in genreNumber:
            genreNumber[genre] = [(i, play)]
        else:
            genreNumber[genre].append((i, play))

    sortedTotalPlays = sorted(totalPlays.items(), key=lambda x: x[1], reverse=True)

    for gen, _ in sortedTotalPlays:
        sortedGenreNumber = sorted(genreNumber[gen], key=lambda x: x[1], reverse=True)[
            :2
        ]

        for num, _ in sortedGenreNumber:
            answer.append(num)

    return answer


result = solution(
    ["classic", "pop", "classic", "classic", "pop"], [500, 600, 150, 800, 2500]
)

File: programmers/lv3/베스트앨범.js
/* 
    lv.3 베스트앨범 

    1.  장르 별 최다 재생 순 정렬
    2.  해당 장르 내에서 재생 내림차순-고유번호 오름차순 정렬
*/

function solution(genres, plays) {
  var answer = [];
  const length = genres.length;

  const genreTotal = {};
  const genrePlays = {};

  for (let i = 0; i < length; i++) {
    const index = i;
    const genre = genres[i];
    const play = plays[i];

    if (!genreTotal[genre]) {
      genreTotal[genre] = play;
    } else {
      genreTotal[genre] += play;
    }

    if (!genrePlays[genre]) {
      genrePlays[genre] = [[play, index]];
    } else {
      genrePlays[genre].push([play, index]);
    }
  }

  for (let genre of genres) {
    genrePlays[genre].sort((prev, curr) => {
      if (prev[0] > curr[0]) return -1;
      if (prev[0] <= curr[0]) return 1;

      if (prev[1] < curr[1]) return -1;
      if (prev[1] >= curr[1]) return 1;
    });
  }

  const sortedGenreTotal = Object.entries(genreTotal)
    .sort(([, a], [, b]) => b - a)
    .map((entry) => entry[0]);

  for (let genre of sortedGenreTotal) {
    const genreTop2 = genrePlays[genre].slice(0, 2);
    const top2Numbers = genreTop2.map((entry) => entry[1]);
    answer = answer.concat(top2Numbers);
  }

  return answer;
}

const genres = ["classic", "pop", "classic", "classic", "pop"];
const plays = [500, 600, 150, 800, 2500];

console.log(solution(genres, plays));
